apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfPublisher.cs

```diff
@@ -0,0 +1,626 @@
+
+        internal HiPerfPublisherImpl(TopicName topicName, Settings settings, Func<Task> shutdown, TaskHelper taskHelper)
+        {
+            TopicName = GaxPreconditions.CheckNotNull(topicName, nameof(topicName));
+            GaxPreconditions.CheckNotNull(settings, nameof(settings));
+            GaxPreconditions.CheckNotNull(settings.Clients, nameof(settings.Clients));
+            settings.Validate();
+            _taskHelper = GaxPreconditions.CheckNotNull(taskHelper, nameof(taskHelper));
+            _shutdown = shutdown;
+
+            // Initialise settings
+            var batchingSettings = settings.BatchingSettings ?? DefaultBatchingSettings();
+            _batchElementCountThreshold = batchingSettings.ElementCountThreshold ?? ApiMaxRequestElementCount;
+            _batchRequestByteThreshold = batchingSettings.RequestByteThreshold ?? ApiMaxRequestByteCount;
+            _batchDelayThreshold = batchingSettings.DelayThreshold ?? TimeSpan.FromHours(1);
+            _scheduler = settings.Scheduler ?? SystemScheduler.Instance;
+            _batchMaxElementCount = settings.BatchMaxElementCount ?? ApiMaxRequestElementCount;
+            _batchMaxByteCount = settings.BatchMaxByteCount ?? ApiMaxRequestByteCount;
+
+            // Initialise internal state
+            _batchesReady = new Queue<Batch>();
+            _idleClients = new Queue<PublisherClient>(settings.Clients);
+            _softStopCts = new CancellationTokenSource();
+            _hardStopCts = new CancellationTokenSource();
+            _shutdownTcs = new TaskCompletionSource<int>();
+        }
+
+        private readonly object _lock = new object();
+        private readonly object _shutdownLock = new object();
+        private readonly IScheduler _scheduler;
+        private readonly TaskHelper _taskHelper;
+        private readonly Func<Task> _shutdown;
+
+        // Batching settings
+        private readonly long _batchElementCountThreshold;
+        private readonly long _batchRequestByteThreshold;
+        private readonly TimeSpan _batchDelayThreshold;
+
+        // Absolute maximum batch values
+        private readonly long _batchMaxElementCount;
+        private readonly long _batchMaxByteCount;
+
+        //Internal state
+        private readonly Queue<PublisherClient> _idleClients;
+        private readonly Queue<Batch> _batchesReady;
+        private Batch _currentBatch;
+        private long _queueElementCount;
+        private long _queueByteCount;
+        private int _batchesInFlightCount;
+        private CancellationTokenSource _softStopCts;
+        private CancellationTokenSource _hardStopCts;
+        private TaskCompletionSource<int> _shutdownTcs;
+        private bool _shutdownStarted;
+
+        /// <inheritdoc/>
+        public override TopicName TopicName { get; }
+
+        /// <inheritdoc/>
+        public override async Task<string> PublishAsync(PubsubMessage message)
+        {
+            CheckShutdown();
```

Comment:
Do we make any guarantees about what happens if we shut down after this check but before a lot of the rest of the method executes?